cranelift/codegen/src/incremental_cache.rs

```diff
@@ -0,0 +1,788 @@
+                            index: *index,
+                        },
+                        func_ref,
+                    ))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        Self {
+            buffer: CachedMachBufferFinalized::new(&mcr.buffer, offset),
+            frame_size: mcr.frame_size,
+            disasm: mcr.disasm.clone(),
+            value_labels_ranges: mcr.value_labels_ranges.clone(),
+            sized_stackslot_offsets: mcr.sized_stackslot_offsets.clone(),
+            dynamic_stackslot_offsets: mcr.dynamic_stackslot_offsets.clone(),
+            bb_starts: mcr.bb_starts.clone(),
+            bb_edges: mcr.bb_edges.clone(),
+            func_refs,
+        }
+    }
+
+    fn expand(self, after: &Function, offset: SourceLoc) -> MachCompileResult {
+        let mut buffer = self.buffer.expand(offset);
+
+        // Adjust external names in relocations.
+        for reloc in buffer.relocs.iter_mut() {
+            if let ExternalName::User {
+                ref mut namespace,
+                ref mut index,
+            } = reloc.name
+            {
+                if let Some(func_ref) = self.func_refs.get(&UserExternalName {
+                    namespace: *namespace,
+                    index: *index,
+                }) {
+                    if let ExternalName::User {
+                        namespace: new_namespace,
+                        index: new_index,
+                    } = &after.dfg.ext_funcs[*func_ref].name
+                    {
+                        *namespace = *new_namespace;
+                        *index = *new_index;
+                    } else {
+                        panic!("unexpected kind of relocation??");
+                    }
+                } else {
+                    panic!("didn't find previous mention of {};{}", namespace, index);
+                }
+            }
+        }
+
+        MachCompileResult {
+            buffer,
+            frame_size: self.frame_size,
+            disasm: self.disasm,
+            value_labels_ranges: self.value_labels_ranges,
+            sized_stackslot_offsets: self.sized_stackslot_offsets,
+            dynamic_stackslot_offsets: self.dynamic_stackslot_offsets,
+            bb_starts: self.bb_starts,
+            bb_edges: self.bb_edges,
+        }
+    }
+}
+
+/// Compute a cache key, and hash it on your behalf.
+///
+/// Since computing the `CacheKey` is a bit expensive, it should be done as least as possible.
+#[inline(never)]
+pub fn compute_cache_key(isa: &dyn TargetIsa, func: &Function) -> (CacheKey, CacheKeyHash) {
+    let srcloc_offset = func
+        .srclocs
+        .get(Inst::new(0))
+        .cloned()
+        .unwrap_or(SourceLoc::new(0));
+
+    let cache_key = CacheKey::new(isa, func, srcloc_offset);
+
+    let hash = {
+        use core::hash::{Hash as _, Hasher as _};
+        let mut hasher = crate::fx::FxHasher::default();
```

Comment:
Shouldn't this use a cryptographically secure hash function? FxHasher is known to have hash collisions. It is a quick and dirty hash function only suitable when you can handle hash collisions, like in a hash table.

Reply:
It doesn't have to be, because right now the incremental cache first looks up using the `CacheKeyHash`, and then compares against the `CacheKey`, exactly as `HashMap`s do first the hashing and then use the `PartialEq` to check for equality.

This was a very safe first way to implement this, and it's also what limits performance of reloads. An alternative way would be to use a safe, longer hash here (as you suggest in this and the other comment above), and see if we can get rid of the equality check in that case. I'd like to experiment with this, likely later.